File: apps/api/src/routers/documents.py
"""Document CRUD endpoints (list, get, patch, delete, reingest).

These endpoints are tenant-scoped via the JWT-or-API-key dependency.
The tenant_id is ALWAYS derived from the auth principal — it is never
read from request body or query string.

Companion to ``routers/ingest.py`` which owns POST /ingest and the
status polling endpoint at /{id}/status.
"""

from __future__ import annotations

import logging
import os
from typing import Annotated, Literal, Optional

from fastapi import APIRouter, Body, Depends, HTTPException, Query, Response, status

from src.core.dependencies import AgentDependencies
from src.core.deps import get_deps
from src.core.tenant import get_tenant_id
from src.models.api import (
    BulkDeleteRequest,
    BulkDeleteResponse,
    DocumentListResponse,
    DocumentRecord,
    DocumentUpdateRequest,
    ReingestResponse,
)
from src.services.ingestion.service import IngestionService

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/v1/documents", tags=["documents"])


def _ingestion_service(
    deps: AgentDependencies = Depends(get_deps),
) -> IngestionService:
    return IngestionService(
        documents_collection=deps.documents_collection,
        chunks_collection=deps.chunks_collection,
    )


def _to_record(doc: dict) -> DocumentRecord:
    """Convert a stored document dict to the API response model.

    Derives ``format`` from the file extension of ``source`` if not stored.
    """
    source = doc.get("source", "") or ""
    fmt = doc.get("format")
    if not fmt:
        ext = os.path.splitext(source)[1].lower().lstrip(".")
        fmt = ext or ""

    return DocumentRecord(
        document_id=str(doc["_id"]),
        title=doc.get("title", ""),
        source=source,
        status=str(doc.get("status", "unknown")),
        chunk_count=int(doc.get("chunk_count", 0) or 0),
        format=fmt,
        size_bytes=doc.get("size_bytes"),
        metadata=doc.get("metadata") or {},
        version=int(doc.get("version", 1) or 1),
        error_message=doc.get("error_message"),
        created_at=doc["created_at"],
        updated_at=doc.get("updated_at", doc["created_at"]),
    )


@router.get("", response_model=DocumentListResponse)
async def list_documents(
    tenant_id: Annotated[str, Depends(get_tenant_id)],
    service: Annotated[IngestionService, Depends(_ingestion_service)],
    page: int = Query(default=1, ge=1, le=10_000),
    page_size: int = Query(default=20, ge=1, le=100),
    status_filter: Optional[Literal["pending", "processing", "ready", "failed"]] = Query(
        default=None, alias="status"
    ),
    search: Optional[str] = Query(default=None, max_length=200),
    sort: Literal["created_at", "updated_at", "title", "status"] = Query(default="created_at"),
    order: Literal["asc", "desc"] = Query(default="desc"),
) -> DocumentListResponse:
    """List documents for the authenticated tenant.

    Pagination is offset-based (page/page_size). Tenant_id is forced from
    the JWT/API-key — it cannot be supplied by the caller.
    """
    items, total = await service.list_documents(
        tenant_id=tenant_id,
        page=page,
        page_size=page_size,
        status=status_filter,
        search=search,
        sort=sort,
        order=order,
    )
    return DocumentListResponse(
        items=[_to_record(d) for d in items],
        total=total,
        page=page,
        page_size=page_size,
    )


@router.delete("", response_model=BulkDeleteResponse)
async def bulk_delete_documents(
    tenant_id: Annotated[str, Depends(get_tenant_id)],
    service: Annotated[IngestionService, Depends(_ingestion_service)],
    body: BulkDeleteRequest = Body(...),
) -> BulkDeleteResponse:
    """Cascade-delete a list of documents (and their chunks) for this tenant.

    Tenant scope is enforced per-id; ids that do not belong to this tenant
    are silently ignored (counted as not deleted) to prevent enumeration.
    """
    deleted = await service.bulk_delete_with_cascade(body.ids, tenant_id)
    logger.info(
        "bulk_delete_documents tenant=%s requested=%d deleted=%d",
        tenant_id,
        len(body.ids),
        deleted,
    )
    return BulkDeleteResponse(requested=len(body.ids), deleted=deleted)


@router.get("/{document_id}", response_model=DocumentRecord)
async def get_document(
    document_id: str,
    tenant_id: Annotated[str, Depends(get_tenant_id)],
    service: Annotated[IngestionService, Depends(_ingestion_service)],
) -> DocumentRecord:
    """Fetch a single document. 404 for missing OR cross-tenant ids."""
    doc = await service.get_document(document_id, tenant_id)
    if not doc:
        raise HTTPException(status_code=404, detail="Document not found")
    return _to_record(doc)


@router.patch("/{document_id}", response_model=DocumentRecord)
async def update_document(
    document_id: str,
    body: DocumentUpdateRequest,
    tenant_id: Annotated[str, Depends(get_tenant_id)],
    service: Annotated[IngestionService, Depends(_ingestion_service)],
) -> DocumentRecord:
    """Update title and/or metadata. Other fields are immutable here."""
    if body.title is None and body.metadata is None:
        raise HTTPException(status_code=422, detail="No fields to update")

    updated = await service.update_metadata(
        document_id=document_id,
        tenant_id=tenant_id,
        title=body.title,
        metadata=body.metadata,
    )
    if not updated:
        raise HTTPException(status_code=404, detail="Document not found")
    return _to_record(updated)


@router.delete("/{document_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_document(
    document_id: str,
    tenant_id: Annotated[str, Depends(get_tenant_id)],
    service: Annotated[IngestionService, Depends(_ingestion_service)],
) -> Response:
    """Cascade-delete a document and all of its chunks.

    Atomic via a Mongo transaction when supported; falls back to a
    sequenced delete (chunks first, then doc) on standalone deployments.
    """
    deleted = await service.delete_document_with_cascade(document_id, tenant_id)
    if not deleted:
        raise HTTPException(status_code=404, detail="Document not found")
    return Response(status_code=status.HTTP_204_NO_CONTENT)


@router.post("/{document_id}/reingest", response_model=ReingestResponse, status_code=202)
async def reingest_document_endpoint(
    document_id: str,
    tenant_id: Annotated[str, Depends(get_tenant_id)],
    service: Annotated[IngestionService, Depends(_ingestion_service)],
) -> ReingestResponse:
    """Mark a document for re-processing.

    Flips status back to ``pending`` so the worker can re-chunk and re-embed
    on its next sweep. Refuses if the document is currently ``processing``
    to avoid racing the existing worker run.
    """
    updated = await service.mark_for_reingestion(document_id, tenant_id)
    if not updated:
        # Could be: missing, wrong tenant, OR already in-flight.
        # We can't distinguish without an extra read — but a second read
        # could leak existence to other tenants, so return 404/409 only
        # after confirming tenant ownership.
        existing = await service.get_document(document_id, tenant_id)
        if not existing:
            raise HTTPException(status_code=404, detail="Document not found")
        raise HTTPException(
            status_code=409,
            detail="Document is currently processing; reingest not allowed",
        )
    return ReingestResponse(document_id=str(updated["_id"]), status=updated["status"])

File: apps/web/lib/api/documents.ts
import { mintBackendToken } from "@/lib/api/token";

const API_URL = process.env.NEXT_PUBLIC_API_URL || "http://localhost:8100";

export type DocumentStatus =
  | "pending"
  | "processing"
  | "ready"
  | "failed"
  | "unknown";

export interface DocumentRecord {
  document_id: string;
  title: string;
  source: string;
  status: DocumentStatus;
  chunk_count: number;
  format: string;
  size_bytes: number | null;
  metadata: Record<string, unknown>;
  version: number;
  error_message: string | null;
  created_at: string;
  updated_at: string;
}

export interface DocumentListResponse {
  items: DocumentRecord[];
  total: number;
  page: number;
  page_size: number;
}

export interface ListDocumentsParams {
  page?: number;
  pageSize?: number;
  status?: DocumentStatus;
  search?: string;
  sort?: "created_at" | "updated_at" | "title" | "status";
  order?: "asc" | "desc";
}

export interface DocumentUpdateInput {
  title?: string;
  metadata?: Record<string, unknown>;
}

export class ApiError extends Error {
  constructor(
    message: string,
    public readonly status: number,
    public readonly code?: string,
  ) {
    super(message);
    this.name = "ApiError";
  }
}

async function authedRequest(
  path: string,
  init: RequestInit = {},
): Promise<Response> {
  const minted = await mintBackendToken();
  if (!minted) throw new ApiError("Not authenticated", 401);

  const headers = new Headers(init.headers);
  headers.set("Authorization", `Bearer ${minted.token}`);
  if (init.body && !(init.body instanceof FormData)) {
    headers.set("Content-Type", "application/json");
  }

  const res = await fetch(`${API_URL}${path}`, {
    ...init,
    headers,
    cache: "no-store",
  });

  if (!res.ok) {
    let message = res.statusText;
    let code: string | undefined;
    try {
      const body = (await res.clone().json()) as {
        detail?: string;
        error?: { code?: string; message?: string };
      };
      message = body.error?.message ?? body.detail ?? message;
      code = body.error?.code;
    } catch {
      // ignore — non-JSON body
    }
    throw new ApiError(message, res.status, code);
  }

  return res;
}

export async function listDocuments(
  params: ListDocumentsParams = {},
): Promise<DocumentListResponse> {
  const search = new URLSearchParams();
  if (params.page) search.set("page", String(params.page));
  if (params.pageSize) search.set("page_size", String(params.pageSize));
  if (params.status) search.set("status", params.status);
  if (params.search) search.set("search", params.search);
  if (params.sort) search.set("sort", params.sort);
  if (params.order) search.set("order", params.order);

  const qs = search.toString();
  const res = await authedRequest(
    `/api/v1/documents${qs ? `?${qs}` : ""}`,
  );
  return (await res.json()) as DocumentListResponse;
}

export async function getDocument(id: string): Promise<DocumentRecord> {
  const res = await authedRequest(
    `/api/v1/documents/${encodeURIComponent(id)}`,
  );
  return (await res.json()) as DocumentRecord;
}

export async function updateDocument(
  id: string,
  input: DocumentUpdateInput,
): Promise<DocumentRecord> {
  const res = await authedRequest(
    `/api/v1/documents/${encodeURIComponent(id)}`,
    { method: "PATCH", body: JSON.stringify(input) },
  );
  return (await res.json()) as DocumentRecord;
}

export async function deleteDocument(id: string): Promise<void> {
  await authedRequest(`/api/v1/documents/${encodeURIComponent(id)}`, {
    method: "DELETE",
  });
}

export async function reingestDocument(id: string): Promise<void> {
  await authedRequest(
    `/api/v1/documents/${encodeURIComponent(id)}/reingest`,
    { method: "POST" },
  );
}

export async function bulkDeleteDocuments(ids: string[]): Promise<void> {
  await authedRequest(`/api/v1/documents`, {
    method: "DELETE",
    body: JSON.stringify({ ids }),
  });
}
